File: enroll_teachers.py
import sqlite3
import os
import json
import base64
import urllib.request
import urllib.parse
import binascii

db_path = os.path.join(os.path.expanduser('~'), 'AppData', 'Roaming', 'school-management-system', 'school.db')
if not os.path.exists(db_path):
    db_path = os.path.join(os.path.expanduser('~'), 'AppData', 'Roaming', 'Tapowan Public School System', 'school.db')

conn = sqlite3.connect(db_path)
cursor = conn.cursor()

try:
    cursor.execute("ALTER TABLE teachers ADD COLUMN photo TEXT")
except:
    pass # Column might already exist

photos_dir = r"C:\Users\Admin\Desktop\School Work\Teachers Photos"
files = [f for f in os.listdir(photos_dir) if f.endswith('.jpg') or f.endswith('.png')]

for file in files:
    full_name = os.path.splitext(file)[0]
    file_path = os.path.join(photos_dir, file)
    print(f"Processing {full_name}...")
    
    with open(file_path, "rb") as f:
        image_data = f.read()
    
    base64_data = "data:image/jpeg;base64," + base64.b64encode(image_data).decode('utf-8')
    
    boundary = '----WebKitFormBoundary7MA4YWxkTrZu0gW'
    body = b'--' + boundary.encode() + b'\r\n'
    body += b'Content-Disposition: form-data; name="file"; filename="photo.jpg"\r\n'
    body += b'Content-Type: image/jpeg\r\n\r\n'
    body += image_data
    body += b'\r\n--' + boundary.encode() + b'--\r\n'
    
    try:
        req = urllib.request.Request('http://127.0.0.1:8000/extract', data=body)
        req.add_header('Content-Type', f'multipart/form-data; boundary={boundary}')
        req.add_header('Content-Length', str(len(body)))
        
        with urllib.request.urlopen(req) as response:
            if response.status != 200:
                print(f"HTTP Error for {full_name}: {response.status}")
                continue
            
            data = json.loads(response.read().decode('utf-8'))
            if data.get('faces') and len(data['faces']) > 0:
                embedding = data['faces'][0]['embedding']
                
                # Role logic
                phone = ""
                role = "Teacher"
                dept = "Staff"
                if full_name == "Rizwan":
                    phone = "9470109644"
                    role = "Administrator"
                    dept = "Management"
                elif full_name == "Raju Gupta":
                    phone = "9546477994"
                    role = "Administrator"
                    dept = "Management"
                
                # Insert/Update Teacher
                cursor.execute("SELECT id FROM teachers WHERE fullName = ?", (full_name,))
                teacher = cursor.fetchone()
                if teacher:
                    cursor.execute("UPDATE teachers SET photo = ?, phone = ? WHERE id = ?", (base64_data, phone, teacher[0]))
                else:
                    cursor.execute("INSERT INTO teachers (fullName, department, phone, photo) VALUES (?, ?, ?, ?)", (full_name, dept, phone, base64_data))
                
                # Insert/Update faceEmbedding
                cursor.execute("DELETE FROM faceEmbeddings WHERE targetType='teachers' AND name=?", (full_name,))
                cursor.execute("""
                    INSERT INTO faceEmbeddings (targetType, name, tag, descriptorJson)
                    VALUES ('teachers', ?, 'teacher', ?)
                """, (full_name, json.dumps(embedding)))
                
                # Insert/Update Users for owners
                if role == "Administrator":
                    username = full_name.replace(" ", "").lower()
                    cursor.execute("SELECT id FROM users WHERE username = ?", (username,))
                    user = cursor.fetchone()
                    if user:
                        cursor.execute("UPDATE users SET role = ?, password = ? WHERE id = ?", (role, "admin123", user[0]))
                    else:
                        cursor.execute("INSERT INTO users (username, fullName, role, password, status) VALUES (?, ?, ?, ?, ?)", (username, full_name, role, "admin123", "Active"))
                    print(f"Added {full_name} as Administrator (Username: {username}, Password: admin123)")
                
                print(f"Enrolled {full_name} successfully.")
            else:
                print(f"No face detected in photo for {full_name}")
    except Exception as e:
        print(f"Error processing {full_name}: {e}")

conn.commit()
conn.close()
print("Finished enrolling teachers!")

File: enroll_teachers.js
const fs = require('fs');
const path = require('path');
const { db } = require('./server/db-sqlite.js'); // Use the existing DB connection

async function enrollTeachers() {
  const photosDir = 'C:\\Users\\Admin\\Desktop\\School Work\\Teachers Photos';
  const files = fs.readdirSync(photosDir).filter(f => f.endsWith('.jpg') || f.endsWith('.png'));
  
  for (const file of files) {
    const fullName = file.replace('.jpg', '').replace('.png', '');
    const filePath = path.join(photosDir, file);
    console.log(`Processing ${fullName}...`);
    
    // Read file and convert to base64
    const fileData = fs.readFileSync(filePath);
    const base64Data = 'data:image/jpeg;base64,' + fileData.toString('base64');
    
    // Send to InsightFace API
    const formData = new FormData();
    formData.append('file', new Blob([fileData], { type: 'image/jpeg' }), 'photo.jpg');
    
    try {
      const response = await fetch('http://127.0.0.1:8000/extract', {
        method: 'POST',
        body: formData
      });
      
      if (!response.ok) {
        console.error(`Failed to extract face for ${fullName}: HTTP ${response.status}`);
        continue;
      }
      
      const data = await response.json();
      if (data.faces && data.faces.length > 0) {
        const embedding = data.faces[0].embedding;
        
        // Ensure teachers table has a photo column
        try {
          db.prepare('ALTER TABLE teachers ADD COLUMN photo TEXT').run();
        } catch(e) { /* Column might already exist */ }
        
        // Insert or update teacher
        const existingTeacher = db.prepare('SELECT id FROM teachers WHERE fullName = ?').get(fullName);
        let phone = "";
        let role = "Teacher";
        
        if (fullName === "Rizwan") { phone = "9470109644"; role = "Administrator"; }
        if (fullName === "Raju Gupta") { phone = "9546477994"; role = "Administrator"; }
        
        if (existingTeacher) {
          db.prepare('UPDATE teachers SET photo = ?, phone = ? WHERE id = ?').run(base64Data, phone, existingTeacher.id);
        } else {
          db.prepare('INSERT INTO teachers (fullName, department, phone, photo) VALUES (?, ?, ?, ?)').run(fullName, role === 'Administrator' ? 'Management' : 'Staff', phone, base64Data);
        }
        
        // Insert or update faceEmbedding
        db.prepare("DELETE FROM faceEmbeddings WHERE targetType='teachers' AND name=?").run(fullName);
        db.prepare("INSERT INTO faceEmbeddings (targetType, name, tag, descriptorJson) VALUES (?, ?, ?, ?)").run('teachers', fullName, 'teacher', JSON.stringify(embedding));
        
        // Add to users if owner
        if (role === "Administrator") {
          const username = fullName.replace(/\s+/g, '').toLowerCase();
          const existingUser = db.prepare('SELECT id FROM users WHERE username = ?').get(username);
          if (existingUser) {
            db.prepare('UPDATE users SET role = ?, password = ? WHERE id = ?').run('Administrator', 'admin123', existingUser.id);
          } else {
            db.prepare('INSERT INTO users (username, fullName, role, password, status) VALUES (?, ?, ?, ?, ?)').run(username, fullName, 'Administrator', 'admin123', 'Active');
          }
          console.log(`Added ${fullName} as Administrator (Username: ${username}, Password: admin123)`);
        }
        
        console.log(`Enrolled ${fullName} successfully.`);
      } else {
        console.log(`No face detected in photo for ${fullName}`);
      }
    } catch (err) {
      console.error(`Error processing ${fullName}:`, err.message);
    }
  }
}

enrollTeachers().then(() => {
  console.log('Finished enrolling teachers!');
  process.exit(0);
});
